packages/core/src/storages/key_value_store.ts

```diff
@@ -273,9 +273,16 @@ export class KeyValueStore {
             return;
         }
 
+        // use half the interval of `persistState` to avoid race conditions
+        const persistStateIntervalMillis = this.config.get('persistStateIntervalMillis')!;
+        const timeoutSecs = persistStateIntervalMillis / 2_000;
+
         this.config.getEventManager().on('persistState', async () => {
             for (const [key, value] of this.cache) {
-                await this.setValue(key, value);
+                await this.setValue(key, value, {
+                    timeoutSecs,
+                    doNotRetryTimeouts: true,
+                });
```

Comment:
Do I understand correctly that the issue was too many concurrent writes _to the same key_, right? 

![image](https://github.com/user-attachments/assets/cb96a1a3-e078-4435-89b3-e37dec372646)

So while this code might still intermingle multiple "epochs" of writes together (see "timeouting" red key-value `cache` pairs from the old `persistState` event and blue writes from the new `persistState` event), it will never issue more than one concurrent write to the same key (because of the `/ 2` timeout)?

Reply:
I see, normally there will be just one key in that cache object (the one for `useState`), so I guess we should be safe. Maybe I should rewrite this to use `Promise.all` and fire all those concurrently, since every entry will have a different key.